{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\" style=\"font-family:'arial';font-size:25px\"> How to calculate prevalence of BRCA1 pathogenic variants carriers in case/control cohorts using Hail </div>\n",
    "\n",
    "**Introduction**\n",
    "\n",
    "This use case is provided by John Baierl and Dr.Paul Pharoah at Cedars-Sinai Medical Center, CA USA. \n",
    "\n",
    "This notebook illustrates the process of determining the prevalence of BRCA1 pathogenic variant carriers within case (ovarian cancer) and control (non-cancer) cohorts using the Allofus (AOU) exome dataset.\n",
    "\n",
    "Within the notebook '05.Genomics_use_case_hail_to_plink.ipynb', we examine the effects of filtering the Hail matrix table (MT) based on chromosome intervals alone versus filtering based on exact locus and alleles.\n",
    "\n",
    "In this notebook, we further demonstrate the effects of these two filtering methods on calculating the prevalence of BRCA1 pathogenic variants.\n",
    "\n",
    "This notebooks uses 4CPU+26G memory and 2/0 workers,runtime is about 4-5mins.\n",
    "\n",
    "**Prerequisite**: Please go through our featured genomic workspace before testing this notebook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "start = datetime.now()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Setup**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import hail as hl\n",
    "hl.init(default_reference = \"GRCh38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket = os.getenv('WORKSPACE_BUCKET')\n",
    "bucket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_wgs_exome_path = os.getenv(\"WGS_EXOME_SPLIT_HAIL_PATH\")\n",
    "mt_wgs_exome_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = hl.read_matrix_table(mt_wgs_exome_path)\n",
    "mt.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_intervals = ['chr17: 43047642-43047643']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt2 = hl.filter_intervals(\n",
    "    mt,\n",
    "    [hl.parse_locus_interval(x,)\n",
    "     for x in test_intervals])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "we can run the cell below to show the shape of data. It means we have 1 row (variant) and 414,830 columns (samples)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt2.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Get GT counts using agg function**\n",
    "\n",
    "We have 1 sample that have 1 alternate allele. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt2.aggregate_entries(hl.agg.counter(mt2.GT.n_alt_alleles()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or we can run the cell below to filter samples that have at least one non-homozygous reference genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter samples that have at least one non-homozygous reference genotype\n",
    "samples_with_alt = mt2.filter_cols(\n",
    "    hl.agg.any(mt2.GT.is_non_ref())\n",
    ")\n",
    "\n",
    "samples_with_alt.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Read BRCA1 pathogenic variants from VAT**\n",
    "\n",
    "Please refer to these two notebooks '03.Genomics_use_case_VAT_Hail' and '04.Genomics_use_case_VAT_bigquery' on how to extract BRCA1 variants from the VAT. We recommned using bigquery to extract variants info from the VAT.\n",
    "\n",
    "Assuming this BRCA1 variant file is saved already in the bucket, and now it can be read back in the cell below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vat_filename = f'{bucket}/data/test/vat6_brca1.tsv'\n",
    "vat_table = hl.import_table(vat_filename,\n",
    "                            impute = True,\n",
    "                            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vat_table.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vat_table.show(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Import genomic data and filter out flagged samples**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start2 = datetime.now()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "auxiliary_path = \"gs://fc-aou-datasets-controlled/v7/wgs/short_read/snpindel/aux\"\n",
    "auxiliary_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = hl.read_matrix_table(mt_wgs_exome_path)\n",
    "\n",
    "# Prepare MT from relatedness, phenotypesm, and ancestry_pred tables\n",
    "relatedness = f'{auxiliary_path}/relatedness'\n",
    "# kin_score_path = f'{relatedness}/relatedness.tsv'\n",
    "related_samples_path = f'{relatedness}/relatedness_flagged_samples.tsv'\n",
    "\n",
    "related_remove = hl.import_table(related_samples_path,\n",
    "                                 types={\"sample_id\":\"tstr\"},\n",
    "                                key=\"sample_id\")\n",
    "mt = mt.anti_join_cols(related_remove)\n",
    "mt.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Keep the case/control samples**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The case/control (1/0) samples are stored in this file \"eoc_all_phenotypes.tsv\" in the bucket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import phenotypes table from bucket as Hail matrix table\n",
    "phenotype_filename = f'{bucket}/data/test/eoc_all_phenotypes.tsv'\n",
    "\n",
    "phenotypes = hl.import_table(phenotype_filename,\n",
    "                            types = {'person_id':hl.tstr},\n",
    "                            impute = True,\n",
    "                            key = 'person_id')\n",
    "\n",
    "phenotypes.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes.show(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**In total there are 735 person_ids in the case and 116259 person_ids in the control cohort**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes.aggregate(hl.agg.counter(phenotypes.is_case))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filters out person_ids not in phenotypes table\n",
    "mt = mt.semi_join_cols(phenotypes)\n",
    "mt = mt.annotate_cols(pheno = phenotypes[mt.s])\n",
    "mt.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter chromosome interval first\n",
    "\n",
    "**Filter chromosome interval first**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create interval column\n",
    "vat_table = vat_table.transmute(position = hl.int32(vat_table.position))\n",
    "vat_table = vat_table.annotate(interval = hl.locus_interval(vat_table.contig, \n",
    "                                                            vat_table.position, vat_table.position + 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt1 = hl.filter_intervals(mt, vat_table.interval.collect())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt1.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After filtering interval, there are 246 variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Calculate carrier prevelance in case/control**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute whether there's at least one heterozygous genotype (is_het) in each individual\n",
    "# in mt1, and annotate the column field 'ind_het' with the result\n",
    "# mt_brca1_burden will have an additional column 'ind_het' indicating whether\n",
    "# there is at least one heterozygous genotype for each individual\n",
    "mt_brca1_burden = mt1.annotate_cols(\n",
    "    ind_het = hl.agg.any(mt1.GT.is_het())\n",
    ")\n",
    "\n",
    "# Drop unwanted columns from mt_brca1_burden. The resulting matrix table (mt_brca1_burden)\n",
    "# will only contain the original columns along with the newly added 'ind_het' column.\n",
    "mt_brca1_burden = mt_brca1_burden.cols()\n",
    "\n",
    "# Group mt_brca1_burden by the 'pheno.is_case' field\n",
    "# Aggregate the number of individuals with at least one LOF (Loss of Function) mutation (n_w_lof) and \n",
    "# the number of individuals without LOF mutations (n_no_lof) within each group.\n",
    "\n",
    "brca1_lof = mt_brca1_burden.group_by(mt_brca1_burden.pheno.is_case).aggregate(\n",
    "    n_w_lof = hl.agg.sum(mt_brca1_burden.ind_het), \n",
    "    n_no_lof = hl.agg.sum(~mt_brca1_burden.ind_het)\n",
    ")\n",
    "\n",
    "# LoF contingency table in case/control\n",
    "brca1_lof.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result is in a 2x2 table. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter exact match on locus+allels\n",
    "\n",
    "**Filter exact match on locus+allels**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat variant strings, parse as variant type to link genomic data\n",
    "vat_table = vat_table.annotate(vid_alt = hl.str('chr') + vat_table.vid.replace('-', ':'))\n",
    "vat_table = vat_table.key_by(**hl.parse_variant(vat_table.vid_alt, reference_genome = 'GRCh38'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotates genomic data with VAT then filters\n",
    "mt2 = mt1.annotate_rows(vat = vat_table[mt1.row_key])\n",
    "mt2 = mt2.semi_join_rows(vat_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt2.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 196 variants left, comparing with 246 variants after interval only filtering. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Calculate carrier prevelance in case/control**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_brca1_burden2 = mt2.annotate_cols(\n",
    "    ind_het = hl.agg.any(mt2.GT.is_het())\n",
    ")\n",
    "\n",
    "mt_brca1_burden2 = mt_brca1_burden2.cols()\n",
    "\n",
    "brca1_lof2 = mt_brca1_burden2.group_by(mt_brca1_burden2.pheno.is_case).aggregate(\n",
    "    n_w_lof = hl.agg.sum(mt_brca1_burden2.ind_het), \n",
    "    n_no_lof = hl.agg.sum(~mt_brca1_burden2.ind_het)\n",
    ")\n",
    "\n",
    "brca1_lof2.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "end = datetime.now()\n",
    "end-start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "**Conclusion**\n",
    "\n",
    "Comparing the two results, a significant difference arises in the frequency of BRCA1 variants within the control group. Specifically, when filtering only by intervals, the frequency is calculated as 0.017% (2090/(2090+114169)), whereas when filtering by exact locus and allele, the frequency is 0.0018% (206/(206+116053)), indicating a difference of about 100x. Such a discrepancy will profoundly impact the comparison of BRCA1 variant frequencies between the case and control groups. Therefore, when utilizing a custom variant list with exact locus and allele information, it is imperative to filter AOU genomic datasets using this precise matched locus and allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "2090/(2090+114169)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "206/(206+116053)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "315.195px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
