{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# How to work with cram files in genomic buckets?\n",
    "\n",
    "**Introduction**\n",
    "\n",
    "This notebook demonstrates how to work directly with CRAM files stored in genomic buckets using new versions of these tools, such as samtools, gatk, bcftools etc., eliminating the need for local downloads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "my_bucket=os.getenv(\"WORKSPACE_BUCKET\")\n",
    "my_bucket"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2",
   "metadata": {},
   "source": [
    "# Where are cram files? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT cat gs://fc-aou-datasets-controlled/v8/wgs/cram/manifest.csv | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "In total, there are 414831-1=414830 cram samples in v8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT cat gs://fc-aou-datasets-controlled/v8/wgs/cram/manifest.csv | head -5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "The v8 cram files include some files from v7/v6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT cat gs://fc-aou-datasets-controlled/v8/wgs/cram/manifest.csv | \\\n",
    "awk -F, 'NR>1 { split($2, path, \"/\"); version=path[7]; count[version]++ } END { for (v in count) print v, count[v] }'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "172120+97460+145250"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "Take a quick at cram files in v8_delta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/** | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11",
   "metadata": {},
   "source": [
    "Check the data size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT du -hs gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": [
    "Check cram files in v7/v6 folder\n",
    "\n",
    "Not all files in v6/v7 are included in v8\n",
    "\n",
    "1832 (1.2%) in v7 and 1130 (1.1%) in v6 are no longer in v8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta/** | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT du -hs gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls gs://fc-aou-datasets-controlled/pooled/wgs/cram/v6_base/** | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT du -hs gs://fc-aou-datasets-controlled/pooled/wgs/cram/v6_base/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18",
   "metadata": {},
   "source": [
    "**A breif summary**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Create the data\n",
    "data = {\n",
    "    \"Version\": [\"v8_delta\", \"v7_delta\", \"v6_base\", \"Total\"],\n",
    "    \"File Count\": [172120, 145250, 97460, 414830],\n",
    "     \"Size (PiB)\": [3.32, 2.62, 1.77 , 7.71]\n",
    "}\n",
    "\n",
    "# Create DataFrame\n",
    "df = pd.DataFrame(data)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_files = 414830\n",
    "total_size_pib = 7.71\n",
    "\n",
    "# Convert PiB to bytes\n",
    "bytes_per_pib = 1024 ** 5\n",
    "total_size_bytes = total_size_pib * bytes_per_pib\n",
    "\n",
    "# Calculate average\n",
    "avg_bytes_per_file = total_size_bytes / total_files\n",
    "\n",
    "# Convert to GiB/GB\n",
    "avg_gib = avg_bytes_per_file / (1024 ** 3)  # 1 GiB = 1024³ bytes\n",
    "avg_gb = avg_bytes_per_file / (1000 ** 3)    # 1 GB  = 1000³ bytes\n",
    "\n",
    "print(f\"Average size per file: {avg_gib:.1f} GiB (binary) or {avg_gb:.2f} GB (SI)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls -l gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/wgs_1000000.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls -l gs://fc-aou-datasets-controlled/pooled/wgs/cram/v6_base/wgs_1000004.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls -l gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta/wgs_1000039.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24",
   "metadata": {},
   "source": [
    "# How to use samtools to check cram files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25",
   "metadata": {},
   "source": [
    "Check the pre-installed samtools version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27",
   "metadata": {},
   "source": [
    "Samtools v1.18 doesn't work for accessing files in a bucket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export ref_fasta='gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta'\n",
    "export file_path='gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/wgs_1000000.cram'\n",
    "\n",
    "samtools view -T $ref_fasta ${file_path} | head -n 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29",
   "metadata": {},
   "source": [
    "**Install a newer version of samtools in order to directly read cram files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget https://github.com/samtools/samtools/releases/download/1.22/samtools-1.22.tar.bz2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bzip2 -d samtools-1.22.tar.bz2\n",
    "!tar -xf samtools-1.22.tar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /home/jupyter/samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd samtools-1.22 # and similarly for bcftools and htslib\n",
    "# here need absolute path\n",
    "./configure --prefix=/home/jupyter/samtools\n",
    "make\n",
    "make install"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34",
   "metadata": {},
   "source": [
    "Check the version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls /home/jupyter/samtools/bin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/samtools/bin\n",
    "./samtools --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export ref_fasta='gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta'\n",
    "export file_path='gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/wgs_1000000.cram'\n",
    "\n",
    "cd /home/jupyter/samtools/bin\n",
    "./samtools view -T $ref_fasta ${file_path} | head -n 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38",
   "metadata": {},
   "source": [
    "## Subset a region to a bam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo $PWD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export ref_fasta=\"gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta\"\n",
    "export out_path=\"${PWD}\"\n",
    "export file_path=\"gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/wgs_1000000.cram\"\n",
    "\n",
    "cd /home/jupyter/samtools/bin\n",
    "./samtools view -b -T $ref_fasta -o ${out_path}/test_subset22.bam ${file_path} chr6:1000000-1005000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls *.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42",
   "metadata": {},
   "source": [
    "we can write output files to bucket direclty."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export ref_fasta=\"gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta\"\n",
    "export out_path=\"${WORKSPACE_BUCKET}/data/test_subset24.bam\"\n",
    "export file_path=\"gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta/wgs_1000059.cram\"\n",
    "\n",
    "cd /home/jupyter/samtools/bin\n",
    "./samtools view -b -T $ref_fasta -o ${out_path} ${file_path} chr6:1000000-1005000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil ls -l ${WORKSPACE_BUCKET}/data/test_subset24.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45",
   "metadata": {},
   "source": [
    "## Check cram file sorting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export file_path=\"gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/wgs_1000000.cram\"\n",
    "\n",
    "cd /home/jupyter/samtools/bin\n",
    "./samtools view -H ${file_path} | grep '^@HD' -A1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47",
   "metadata": {},
   "source": [
    "OR we can use gsutil+preinstalled samtools to stream the header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT cat gs://fc-aou-datasets-controlled/pooled/wgs/cram/v8_delta/wgs_1000000.cram | samtools view -H - | grep '^@HD' -A1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49",
   "metadata": {},
   "source": [
    "# How to use gatk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gatk PrintReads  \\\n",
    "        -I gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta/wgs_1000039.cram \\\n",
    "        -L \"chr2:88750000-90235368\" \\\n",
    "        -R gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta \\\n",
    "        -O igk.cram \\\n",
    "        --gcs-project-for-requester-pays ${GOOGLE_PROJECT} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls igk.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52",
   "metadata": {},
   "source": [
    "**Example 2: using gatk on vcf files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls -l gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/clinvar/vcf/0000000001.* |head -5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "export ref_fasta=\"gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta\"\n",
    "export out_path=\"${WORKSPACE_BUCKET}/data/region_chr1.vcf.bgz\"\n",
    "export file_path=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/clinvar/vcf/0000000001.vcf.bgz\"\n",
    "\n",
    "gatk SelectVariants \\\n",
    "  -R ${ref_fasta} \\\n",
    "  -V ${file_path} \\\n",
    "  -O ${out_path} \\\n",
    "  -L chr1:1000000-1005000 \\\n",
    "  --gcs-project-for-requester-pays ${GOOGLE_PROJECT} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil ls -l ${WORKSPACE_BUCKET}/data/region_chr1.vcf.bgz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56",
   "metadata": {},
   "source": [
    "# How to use samtools via dsub?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env USER_NAME={USER_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash --out test_ID\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "ref_path=\"gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta\"\n",
    "file_path=\"gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta/wgs_1000059.cram\"\n",
    "\n",
    "aou_dsub \\\n",
    "  --name \"${JOB_NAME}\" \\\n",
    "  --logging \"${WORKSPACE_BUCKET}/data/logging\" \\\n",
    "  --image conormesser/splash:v2.6.1 \\\n",
    "  --boot-disk-size 20 \\\n",
    "  --disk-size 200 \\\n",
    "  --env GOOGLE_PROJECT=\"${GOOGLE_PROJECT}\" \\\n",
    "  --env ref_fasta=\"${ref_path}\" \\\n",
    "  --env cram_file=\"${file_path}\" \\\n",
    "  --output-recursive out_path=\"${WORKSPACE_BUCKET}/data/cram_result/\" \\\n",
    "  --command 'export GCS_OAUTH_TOKEN=`gcloud auth print-access-token` && \\\n",
    "             export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT` && \\\n",
    "             echo $GCS_OAUTH_TOKEN > ${out_path}/GCS_OAUTH_TOKEN.txt && \\\n",
    "             echo $GCS_REQUESTER_PAYS_PROJECT > ${out_path}/GCS_REQUESTER_PAYS_PROJECT.txt && \\\n",
    "             echo $GOOGLE_PROJECT > ${out_path}/GOOGLE_PROJECT2.txt && \\\n",
    "             echo ${WORKSPACE_BUCKET} > ${out_path}/WORKSPACE_BUCKET.txt && \\\n",
    "             samtools view -b -T ${ref_fasta} -F 1036 -e \"sclen <= 30\" -o ${out_path}/test_subset3.bam ${cram_file} chr6:1000000-1005000'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save this Python variable value as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_ID={test_ID}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "dstat \\\n",
    "    --provider google-batch \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --location us-central1 \\\n",
    "    --jobs \"${JOB_ID}\" \\\n",
    "    --users \"${USER_NAME}\" \\\n",
    "    --status '*'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gsutil ls -l ${WORKSPACE_BUCKET}/data/cram_result/test_subset3.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62",
   "metadata": {},
   "source": [
    "# How to use bcftools?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63",
   "metadata": {},
   "source": [
    "Install a new version of bcftools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget https://github.com/samtools/bcftools/releases/download/1.22/bcftools-1.22.tar.bz2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65",
   "metadata": {},
   "outputs": [],
   "source": [
    "!bzip2 -d bcftools-1.22.tar.bz2\n",
    "!tar -xf bcftools-1.22.tar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /home/jupyter/bcftools/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd bcftools-1.22  # and similarly for bcftools and htslib\n",
    "./configure --prefix=/home/jupyter/bcftools/\n",
    "make\n",
    "make install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/bcftools/bin\n",
    "./bcftools --version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69",
   "metadata": {},
   "source": [
    "**Check vcf files using bcftools**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export ref_fasta='gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta'\n",
    "export out_path=\"${PWD}\"\n",
    "export file_path='gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/clinvar/vcf'\n",
    "\n",
    "cd /home/jupyter/bcftools/bin\n",
    "./bcftools query -f '%CHROM\\t%POS\\t%REF\\t%ALT\\n' ${file_path}/0000000001.vcf.bgz | head -5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71",
   "metadata": {},
   "source": [
    "**Check cram files using bcftools**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "export GCS_OAUTH_TOKEN=`gcloud auth print-access-token`\n",
    "export GCS_REQUESTER_PAYS_PROJECT=`echo $GOOGLE_PROJECT`\n",
    "export ref_fasta='gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta'\n",
    "export out_path=\"${PWD}\"\n",
    "export file_path='gs://fc-aou-datasets-controlled/pooled/wgs/cram/v7_delta'\n",
    "\n",
    "cd /home/jupyter/bcftools/bin\n",
    "./bcftools mpileup -f $ref_fasta -r chr6:1000000-1005000 ${file_path}/wgs_1000039.cram \\\n",
    "| ./bcftools call -mv -Oz -o ${out_path}/chr6.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls -l chr6.vcf.gz"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "372px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
